src/SourceBuild/content/eng/tools/tasks/Microsoft.DotNet.UnifiedBuild.Tasks/WriteSbrpUsageReport.cs

```diff
@@ -0,0 +1,195 @@
+                .Select(pkg => pkg)
+                .ToArray();
+
+            foreach (PackageInfo sbrpPkg in _sbrpPackages.Values)
+            {
+                foreach (PackageInfo unrefPkg in unrefPkgs)
+                {
+                    var unref = sbrpPkg.References.Keys
+                        .SingleOrDefault(path => path.Contains(SbrpRepoName) && path.Contains($"{unrefPkg.Name}.{unrefPkg.Version}"));
+                    if (unref != null)
+                    {
+                        Log.LogMessage($"Removing {unrefPkg.Id} from {sbrpPkg.Id}'s references.");
+                        sbrpPkg.References.Remove(unref);
+                        hasPurged = true;
+                    }
+                }
+            }
+        } while (hasPurged);
+    }
+
+    private IEnumerable<PackageInfo> GetUnreferencedSbrps() =>
+        _sbrpPackages.Values.Where(pkg => pkg.References.Count == 0);
+
+    private string GetSbrpPackagesPath(string packageType) =>
+        Path.Combine(SrcPath, SbrpRepoName, "src", packageType, "src");
+
+    private void ReadSbrpPackages(string packageType, bool trackTfms)
+    {
+        EnumerationOptions options = new() { RecurseSubdirectories = true };
+
+        foreach (string projectPath in Directory.GetFiles(GetSbrpPackagesPath(packageType), "*.csproj", options))
+        {
+            DirectoryInfo directory = Directory.GetParent(projectPath);
+            string version = directory.Name;
+            string projectName = Path.GetFileNameWithoutExtension(projectPath);
+            PackageInfo info = new()
+            {
+                Version = version,
+                Name = projectName[..(projectName.Length - 1 - version.Length)],
+                Path = directory.FullName,
+            };
+
+            if (trackTfms)
+            {
+                XDocument xmlDoc = XDocument.Load(projectPath);
+                // Reference packages are generated using the TargetFrameworks property
+                // so there is no need to handle the TargetFramework property.
+                string[] tfms = xmlDoc.Element("Project")?
+                    .Elements("PropertyGroup")
+                    .Elements("TargetFrameworks")
+                    .FirstOrDefault()?.Value?.Split(';');
+
+                if (tfms == null || tfms.Length == 0)
+                {
+                    Log.LogError($"No TargetFrameworks were delected in {projectPath}.");
+                }
+
+                info.Tfms = new HashSet<string>(tfms);
+            }
+            else
+            {
+                info.Tfms = [];
+            }
+
+            _sbrpPackages.Add(info.Id, info);
+            Log.LogMessage($"Detected package: {info.Id}");
+        }
+    }
+
+    private void ScanProjectReferences()
+    {
+        EnumerationOptions options = new() { RecurseSubdirectories = true };
+
+        foreach (string projectJsonFile in Directory.GetFiles(SrcPath, "project.assets.json", options))
```

Comment:
I wonder if there's a more performant way of doing this.

Also, we already have a similar logic of finding project.assets.json files in the VMR: https://github.com/dotnet/dotnet/blob/c21cd606d4c668767a4c55b8de8822367ab686b3/repo-projects/Directory.Build.targets#L692-L698

Would it be possible to consolidate that somehow?

Reply:
> Also, we already have a similar logic of finding project.assets.json files in the VMR:

The actual implementation resides at: https://github.com/dotnet/dotnet/blob/main/eng/tools/tasks/Microsoft.DotNet.UnifiedBuild.Tasks/UsageReport/WritePackageUsageData.cs#L141

The core logic is identical - Directory.GetFiles(RootDir, "project.assets.json", SearchOption.AllDirectories)

I can certainly share the implementation.

Reply:
I'm not sure if it's worth it. Just wanted to mention it. I think what you have is OK.